### PR TITLE
Improve enabledFlat check when there are a lot of categories

### LINKED ISSUE #37422: [Issue] Improve enabledFlat check when there are a lot of categories
This issue is automatically created based on existing pull request: magento/magento2#37185: Improve enabledFlat check when there are a lot of categories

---------
For the menu block the call to `if ($this->flatState->isAvailable()) {` can cost up to 20ms if more then 1k of categories are loaded. This can be optimized by keeping track of the isAvailable return value depending on the used storeId.

A similar approach is done in `vendor/magento/module-catalog/Model/ResourceModel/Product/Collection.php` 

This is a minor improvement, but still worthwhile in my opinion.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by igorwulff**
For the menu block the call to `if ($this->flatState->isAvailable()) {` can cost up to 20ms if more then 1k of categories are loaded. This can be optimized by keeping track of the isAvailable return value depending on the used storeId.

A similar approach is done in `vendor/magento/module-catalog/Model/ResourceModel/Product/Collection.php` 

This is a minor improvement, but still worthwhile in my opinion.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37422: Improve enabledFlat check when there are a lot of categories

### CODE REVIEW

**Comment by igorwulff** [comment]
@magento run all tests

**Comment by igorwulff** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Model/Category.php [hunk 1]
```diff
@@ -288,14 +293,22 @@ public function __construct(
...
-        if ($this->flatState->isAvailable()) {
+        if ($this->isEnabledFlat()) {
             $this->_init(\Magento\Catalog\Model\ResourceModel\Category\Flat::class);
             $this->_useFlatResource = true;
         } else {
             $this->_init(\Magento\Catalog\Model\ResourceModel\Category::class);
         }
     }
 
+    private function isEnabledFlat()
```
Comment (leonhelmus) [inline app/code/Magento/Catalog/Model/Category.php]: Missing strict type return argument.

File: app/code/Magento/Catalog/Model/Category.php [hunk 2]
```diff
@@ -208,6 +208,11 @@ class Category extends \Magento\Catalog\Model\AbstractModel implements
      */
     protected $metadataService;
 
+    /**
+     * @var array
+     */
+    private static $flatEnabled = [];
```
Comment (leonhelmus) [inline app/code/Magento/Catalog/Model/Category.php]: please strict type this class property.

**Review by leonhelmus — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/Category.php (hunk 2 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Category.php]:
```suggestion
    private static array $flatEnabled = [];
```

File: app/code/Magento/Catalog/Model/Category.php (hunk 1 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Category.php]:
```suggestion
    private function isEnabledFlat(): bool
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please review and fix failing tests

**Comment by engcom-Delta** [comment]
@magento run all tests

**Comment by engcom-Delta** [comment]
@magento run  all tests

File: app/code/Magento/Catalog/Model/Category.php [hunk 3]
```diff
@@ -109,6 +109,7 @@ class Category extends \Magento\Catalog\Model\AbstractModel implements
      *
      * @var \Magento\UrlRewrite\Model\UrlRewrite
      * @deprecated 102.0.0
+     * @see 227208e9
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Category.php]: @engcom-Delta what is it? Do you feed everyone will understand it?

**Comment by engcom-Delta** [comment]
@magento run all tests

**Comment by engcom-Delta** [comment]
@magento run Integration Tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @igorwulff,

Thanks for the contribution and collaboration!

The code changes seem fine for us, but could you please check the failed tests.

Thanks

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Integration Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Integration Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Integration Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Integration Tests

**Comment by engcom-Charlie** [comment]
@magento run Unit Tests, Integration Tests

**Comment by engcom-Charlie** [comment]
@magento run Unit Tests, Integration Tests

**Comment by engcom-Charlie** [comment]
Hi @igorwulff,

I have fixed few unit test and integration test failures. 

There are still few remaining from unit as well as integration tests. Can you please have a look into them in order to move this PR further.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Pull request closed without merging**